File: tools/diag_handler.js
// ============================================================
// 天音设备 — 传送门handler诊断脚本
// 目的: 抓取C++ handler被调用时的参数，验证能否直接调用跳图
// ============================================================

var libc = Process.getModuleByName("libc.so");

// === 解析函数地址 ===
// 方法1: 从libtestcpp.so基址+偏移计算 (推荐)
// 方法2: 使用之前验证过的绝对地址 (fallback)
var handlerAddr, func1Addr, func2Addr;
var ABS_HANDLER = ptr(0xc276018);
var ABS_FUNC1   = ptr(0xc354228);
var ABS_FUNC2   = ptr(0xc35442c);
var REF_BASE    = ptr(0xc19c000); // 绝对地址对应的参考基址

try {
    var libtestcpp = Process.getModuleByName("libtestcpp.so");
    var base = libtestcpp.base;
    handlerAddr = base.add(ABS_HANDLER.sub(REF_BASE));
    func1Addr   = base.add(ABS_FUNC1.sub(REF_BASE));
    func2Addr   = base.add(ABS_FUNC2.sub(REF_BASE));
    send({t: 'info', msg: '方式1: 从libtestcpp.so基址计算'});
    send({t: 'info', msg: '  base: ' + base});
} catch(e) {
    send({t: 'warn', msg: 'libtestcpp.so 未找到，尝试搜索模块...'});
    // 枚举模块找到包含handler地址的那个
    var found = false;
    Process.enumerateModules().forEach(function(m) {
        if (found) return;
        if (ABS_HANDLER.compare(m.base) >= 0 && ABS_HANDLER.compare(m.base.add(m.size)) < 0) {
            var b = m.base;
            handlerAddr = b.add(ABS_HANDLER.sub(REF_BASE));
            func1Addr   = b.add(ABS_FUNC1.sub(REF_BASE));
            func2Addr   = b.add(ABS_FUNC2.sub(REF_BASE));
            send({t: 'info', msg: '方式2: 从模块 ' + m.name + ' 计算'});
            send({t: 'info', msg: '  base: ' + b + ' size: ' + m.size});
            found = true;
        }
    });
    if (!found) {
        send({t: 'warn', msg: '未找到匹配模块，使用绝对地址'});
        handlerAddr = ABS_HANDLER;
        func1Addr   = ABS_FUNC1;
        func2Addr   = ABS_FUNC2;
    }
}

send({t: 'info', msg: 'handler: ' + handlerAddr});
send({t: 'info', msg: 'func1:   ' + func1Addr});
send({t: 'info', msg: 'func2:   ' + func2Addr});

// === Step 1: 找游戏socket fd ===
var gameFd = -1;
var sessionKey = 0;
var portalCallCount = 0;
var capturedParams = [];  // 保存每次handler调用的参数

function findGameFd() {
    var getpeernamePtr = libc.getExportByName("getpeername");
    if (!getpeernamePtr) return -1;
    var getpeername = new NativeFunction(getpeernamePtr, "int", ["int", "pointer", "pointer"]);

    for (var fd = 30; fd <= 200; fd++) {
        try {
            var addr = Memory.alloc(128);
            var addrLen = Memory.alloc(4);
            addrLen.writeInt(128);
            if (getpeername(fd, addr, addrLen) === 0) {
                var family = addr.readU16();
                if (family === 2) { // AF_INET
                    var port = ((addr.add(2).readU8() << 8) | addr.add(3).readU8());
                    if (port === 30002) {
                        send({t: 'fd_found', fd: fd});
                        return fd;
                    }
                }
            }
        } catch (e) {}
    }
    return -1;
}

gameFd = findGameFd();
if (gameFd < 0) {
    send({t: 'warn', msg: '未找到游戏socket fd=30002，将从send调用中自动检测'});
}

function isGameFd(fd) {
    if (gameFd > 0) return fd === gameFd;
    return false;
}

// === Step 2: Hook send — 抓传送包 + 自动检测fd和key ===
Interceptor.attach(libc.getExportByName("send"), {
    onEnter: function(args) {
        var fd = args[0].toInt32();
        var buf = args[1];
        var len = args[2].toInt32();

        // 自动检测游戏fd
        if (gameFd < 0 && len >= 10 && len <= 500) {
            gameFd = fd;
            send({t: 'fd_found', fd: fd, msg: '从send()自动检测到游戏fd'});
        }
        if (!isGameFd(fd)) return;

        // 自动检测XOR密钥 (17B心跳包 byte0=0x01)
        if (sessionKey === 0 && len === 17) {
            var tryKey = buf.readU8() ^ 0x01;
            if (tryKey > 0 && tryKey < 256) {
                sessionKey = tryKey;
                send({t: 'key_found', key: sessionKey, hex: '0x' + tryKey.toString(16)});
            }
        }

        // 检测传送包: 29B, type=0x03 (plain)
        if (len === 29 && sessionKey > 0) {
            var ptype = buf.readU8() ^ sessionKey;
            if (ptype === 0x03) {
                var plain = '';
                for (var i = 0; i < len; i++) {
                    plain += ('0' + (buf.add(i).readU8() ^ sessionKey).toString(16)).slice(-2);
                }
                send({t: 'portal_packet', plain: plain, callIndex: portalCallCount,
                    msg: '=== 传送包已发送 (' + len + 'B) ==='});

                // 详细字节分解
                var detail = 'Byte分解:\n';
                for (var j = 0; j < len; j++) {
                    var b = parseInt(plain.substring(j*2, j*2+2), 16);
                    detail += '  [' + j + ']=0x' + ('0' + b.toString(16)).slice(-2) + ' (' + b + ')\n';
                }
                send({t: 'portal_detail', detail: detail});
            }
        }
    }
});

// === Step 3: Hook recv — 自动检测key (备用) ===
Interceptor.attach(libc.getExportByName("recv"), {
    onEnter: function(args) {
        this.fd = args[0].toInt32();
        this.buf = args[1];
        this.doProcess = isGameFd(this.fd);
    },
    onLeave: function(ret) {
        if (!this.doProcess) return;
        var realLen = ret.toInt32();
        if (realLen <= 4) return;

        if (sessionKey === 0) {
            var b0 = this.buf.readU8();
            var b1 = this.buf.add(1).readU8();
            var b2 = this.buf.add(2).readU8();
            var b3 = this.buf.add(3).readU8();
            if (b1 === b2 && b2 === b3) {
                var tryKey = b0 ^ 0x02;
                if (tryKey > 0 && tryKey < 256) {
                    sessionKey = tryKey;
                    send({t: 'key_found', key: sessionKey, side: 'recv', hex: '0x' + tryKey.toString(16)});
                }
            }
        }
    }
});

// === Step 4: 反汇编handler函数 (前60条指令) ===
function disasmAt(addr, count, label) {
    var result = [];
    var cursor = addr;
    for (var i = 0; i < count; i++) {
        try {
            var insn = Instruction.parse(cursor);
            var bytes = '';
            var arr = new Uint8Array(cursor.readByteArray(insn.size));
            for (var j = 0; j < insn.size; j++) {
                bytes += ('0' + arr[j].toString(16)).slice(-2);
            }
            result.push({
                offset: cursor.sub(addr).toInt32(),
                addr: cursor.toString(),
                bytes: bytes,
                asm: insn.mnemonic + ' ' + insn.operands
            });
            cursor = cursor.add(insn.size);
        } catch(e) {
            result.push({offset: cursor.sub(addr).toInt32(), err: e.toString()});
            cursor = cursor.add(2);
        }
    }
    send({t: 'disasm', label: label, addr: addr.toString(), instructions: result});
    return result;
}

// === Step 5: Hook handler — 核心！抓取调用参数 ===
try {
    Interceptor.attach(handlerAddr, {
        onEnter: function(args) {
            portalCallCount++;
            var params = {
                count: portalCallCount,
                r0: args[0].toString(),
                r1: args[1].toString(),
                r2: args[2].toString(),
                r3: args[3].toString(),
                lr: this.context.lr.toString(),
                sp: this.context.sp.toString(),
                time: Date.now()
            };

            // 尝试读字符串参数 (URL等)
            // 试args[1] (可能是URL指针的指针)
            try {
                if (args[1] && !args[1].isNull()) {
                    // 可能是指向指针的指针
                    var ptr1 = args[1].readPointer();
                    if (ptr1 && !ptr1.isNull()) {
                        var str1 = ptr1.readCString();
                        if (str1 && str1.length > 0 && str1.length < 500) {
                            params.url = str1;
                            send({t: 'handler_url', url: str1, count: portalCallCount,
                                msg: '>>> 发现URL: ' + str1});
                        }
                    }
                    // 也可能是直接字符串
                    var str1d = args[1].readCString();
                    if (str1d && str1d.length > 0 && str1d.length < 500 && !params.url) {
                        params.url = str1d;
                        send({t: 'handler_url', url: str1d, count: portalCallCount,
                            msg: '>>> 发现URL(direct): ' + str1d});
                    }
                }
            } catch(e) {}

            // 试args[0] (handlerObj — 可能包含map info)
            try {
                if (args[0] && !args[0].isNull()) {
                    // 读前32字节看结构
                    var dump = '';
                    for (var di = 0; di < 32; di++) {
                        dump += ('0' + args[0].add(di).readU8().toString(16)).slice(-2);
                    }
                    params.objDump = dump;
                }
            } catch(e) {}

            // 试args[2]
            try {
                if (args[2] && !args[2].isNull()) {
                    var str2 = args[2].readCString();
                    if (str2 && str2.length > 0 && str2.length < 500) {
                        params.arg2Str = str2;
                        send({t: 'handler_str2', str: str2, count: portalCallCount});
                    }
                }
            } catch(e) {}

            capturedParams.push(params);

            send({t: 'handler_enter',
                count: portalCallCount,
                r0: params.r0,
                r1: params.r1,
                r2: params.r2,
                r3: params.r3,
                lr: params.lr,
                url: params.url || '(未识别)',
                objDump: params.objDump || '(无)',
                msg: '=== HANDLER 被调用! #' + portalCallCount + ' ==='
            });
        },
        onLeave: function(ret) {
            send({t: 'handler_leave', count: portalCallCount, ret: ret.toString()});
        }
    });
    send({t: 'info', msg: '[OK] handler hook 已安装'});
} catch(e) {
    send({t: 'error', msg: 'handler hook 失败: ' + e});
}

// === Step 6: Hook func1 (初始化函数) ===
try {
    Interceptor.attach(func1Addr, {
        onEnter: function(args) {
            send({t: 'func1_enter',
                r0: args[0].toString(), r1: args[1].toString(),
                r2: args[2].toString(), r3: args[3].toString(),
                msg: '--- func1 被调用 ---'
            });
        },
        onLeave: function(ret) {
            send({t: 'func1_leave', ret: ret.toString()});
        }
    });
    send({t: 'info', msg: '[OK] func1 hook 已安装'});
} catch(e) {
    send({t: 'error', msg: 'func1 hook 失败: ' + e});
}

// === Step 7: Hook func2 (构造包) ===
try {
    Interceptor.attach(func2Addr, {
        onEnter: function(args) {
            send({t: 'func2_enter',
                r0: args[0].toString(), r1: args[1].toString(),
                r2: args[2].toString(), r3: args[3].toString(),
                msg: '--- func2 被调用 ---'
            });
        },
        onLeave: function(ret) {
            send({t: 'func2_leave', ret: ret.toString()});
        }
    });
    send({t: 'info', msg: '[OK] func2 hook 已安装'});
} catch(e) {
    send({t: 'error', msg: 'func2 hook 失败: ' + e});
}

// === Step 8: 反汇编handler (了解内部逻辑) ===
disasmAt(handlerAddr, 50, 'handler');
disasmAt(func1Addr, 40, 'func1');
disasmAt(func2Addr, 40, 'func2');

// === RPC — 供Python端调用 ===
rpc.exports = {
    // 获取当前状态
    getStatus: function() {
        return JSON.stringify({
            gameFd: gameFd,
            sessionKey: sessionKey,
            portalCallCount: portalCallCount,
            handlerAddr: handlerAddr.toString(),
            func1Addr: func1Addr.toString(),
            func2Addr: func2Addr.toString(),
            capturedCount: capturedParams.length
        });
    },

    // 获取最近一次handler调用的参数
    getLastCall: function() {
        if (capturedParams.length === 0) return JSON.stringify({error: '还没有handler被调用'});
        return JSON.stringify(capturedParams[capturedParams.length - 1]);
    },

    // 获取所有handler调用记录
    getAllCalls: function() {
        return JSON.stringify(capturedParams);
    },

    // 尝试直接调用handler — 用自定义URL跳图
    tryCallHandler: function(urlStr) {
        if (!urlStr) return '用法: tryCallHandler("xqj://map?name=目标地图")';
        if (capturedParams.length === 0) return '错误: 还没有抓取到handlerObj，请先走一次传送门';

        // 用上次抓到的handlerObj (R0)
        var lastCall = capturedParams[capturedParams.length - 1];
        var handlerObj = ptr(lastCall.r0);

        // 构造URL
        var url = Memory.allocUtf8String(urlStr);
        var urlPtr = Memory.alloc(4);
        urlPtr.writePointer(url);

        // 调用handler
        try {
            var handlerFn = new NativeFunction(handlerAddr, 'void', ['pointer', 'pointer']);
            handlerFn(handlerObj, urlPtr);
            return '已调用 handler("' + urlStr + '"), handlerObj=' + handlerObj;
        } catch(e) {
            return '调用失败: ' + e;
        }
    },

    // 尝试调用 — 完整模拟 (先用func1初始化，再用handler)
    tryCallFull: function(urlStr) {
        if (!urlStr) return '用法: tryCallFull("xqj://map?name=目标地图")';

        // func1 可能是创建初始化结构的
        // 分配一些内存给handlerObj
        var handlerObj = Memory.alloc(256);
        var url = Memory.allocUtf8String(urlStr);
        var urlPtr = Memory.alloc(4);
        urlPtr.writePointer(url);

        try {
            // 先调func1初始化
            var func1Fn = new NativeFunction(func1Addr, 'pointer', ['pointer', 'pointer']);
            var initRet = func1Fn(handlerObj, urlPtr);
            send({t: 'try_call', msg: 'func1 返回: ' + initRet});

            // 再调handler
            var handlerFn = new NativeFunction(handlerAddr, 'void', ['pointer', 'pointer']);
            handlerFn(handlerObj, urlPtr);

            return '已调用 func1 + handler("' + urlStr + '")';
        } catch(e) {
            return '调用失败: ' + e;
        }
    },

    // 获取密钥
    getKey: function() { return sessionKey; }
};

// === 完成 ===
send({t: 'ready',
    msg: '========================================',
    msg2: '诊断脚本就绪！请操作：',
    msg3: '',
    msg4: '【步骤1】在游戏中点击任意传送门走一次',
    msg5: '   -> 观察控制台输出：handler_enter / func1 / func2 / portal_packet',
    msg6: '',
    msg7: '【步骤2】走完后看控制台，确认抓到了：',
    msg8: '   - handler的URL参数 (handler_url)',
    msg9: '   - 传送包明文 (portal_packet)',
    msg10: '   - 函数调用顺序',
    msg11: '',
    msg12: '【步骤3】然后运行 tryCallHandler("xqj://map?name=地图名") 尝试直接跳图',
    msg13: '========================================',
    gameFd: gameFd,
    key: sessionKey
});

File: tools/diag_handler.py
"""
天音设备 — 传送门handler诊断工具
======================================
目的: 抓取C++ handler参数 → 验证能否直接调用跳图函数

使用方法:
  1. 确保天音设备已连接 (ADB 127.0.0.1:16384, Frida 127.0.0.1:27056)
  2. 确保游戏已登录，角色站在任意位置
  3. 运行此脚本: D:/Setup_and_Downloads/Setup/python3.12.4/python.exe tools/diag_handler.py
  4. 按控制台提示操作
"""
import frida, time, sys, json

FILES = "E:/DATA/xunqinol_script/tools"
FRIDA_HOST = "127.0.0.1:27056"

def on_msg(msg, data):
    p = msg.get("payload", {})
    if not isinstance(p, dict):
        return
    t = p.get("t", "")

    if t == "ready":
        print(f"\n{'='*60}")
        print(f"[就绪] {p.get('msg', '')}")
        for k in [f'msg{i}' for i in range(2, 14)]:
            v = p.get(k)
            if v:
                print(v)
        print(f"fd={p.get('gameFd')} key={p.get('key')}")
        print(f"{'='*60}\n")

    elif t == "info":
        print(f" [信息] {p['msg']}")

    elif t == "warn":
        print(f" [警告] {p['msg']}")

    elif t == "error":
        print(f" [错误] {p['msg']}")

    elif t == "fd_found":
        print(f" [FD] 游戏socket fd={p.get('fd')} {p.get('msg','')}")

    elif t == "key_found":
        print(f" [密钥] XOR key = 0x{p['key']:02x} ({p.get('side','send')}侧检测)")

    elif t == "handler_enter":
        print(f"\n{'*'*50}")
        print(f"[HANDLER] 第{p['count']}次调用 ━━━━━━━━━━━━━━━━")
        print(f"  R0 (handlerObj): {p['r0']}")
        print(f"  R1 (urlPtr):     {p['r1']}")
        print(f"  R2:              {p['r2']}")
        print(f"  R3:              {p['r3']}")
        print(f"  LR (返回地址):   {p['lr']}")
        print(f"  URL:             {p.get('url', '?')}")
        print(f"  objDump(32B):    {p.get('objDump', '?')}")
        print(f"{'*'*50}")

    elif t == "handler_url":
        print(f"  >>> URL字符串: {p['url']}")

    elif t == "handler_str2":
        print(f"  >>> R2字符串: {p['str']}")

    elif t == "handler_leave":
        print(f"  <<< handler 返回: {p['ret']}")

    elif t == "func1_enter":
        print(f"  [func1] ENTER R0={p['r0']} R1={p['r1']} R2={p['r2']} R3={p['r3']}")

    elif t == "func1_leave":
        print(f"  [func1] LEAVE ret={p['ret']}")

    elif t == "func2_enter":
        print(f"  [func2] ENTER R0={p['r0']} R1={p['r1']} R2={p['r2']} R3={p['r3']}")

    elif t == "func2_leave":
        print(f"  [func2] LEAVE ret={p['ret']}")

    elif t == "portal_packet":
        print(f"\n  [传送包] 明文({len(p['plain'])//2}B): {p['plain']}")

    elif t == "portal_detail":
        print(p['detail'])

    elif t == "disasm":
        label = p['label']
        insts = p.get('instructions', [])
        print(f"\n  === 反汇编 {label} @ {p['addr']} ===")
        for ins in insts:
            if 'err' in ins:
                print(f"    {ins['offset']:4d}  [ERR] {ins['err']}")
            else:
                print(f"    {ins['offset']:4d}  {ins['bytes']:<12s}  {ins['asm']}")

    elif t == "try_call":
        print(f"  [尝试调用] {p['msg']}")

    else:
        # 忽略过于频繁的消息
        pass

    sys.stdout.flush()


def main():
    print("=" * 60)
    print("  天音 — 传送门handler诊断工具")
    print("=" * 60)

    # 读取JS
    js_path = f"{FILES}/diag_handler.js"
    with open(js_path, "r", encoding="utf-8") as f:
        JS = f.read()

    # 连接天音
    print(f"\n连接 Frida: {FRIDA_HOST} ...")
    try:
        dev = frida.get_device_manager().add_remote_device(FRIDA_HOST)
    except Exception as e:
        print(f"[失败] 无法连接Frida: {e}")
        print("请确认:")
        print("  1. 天音模拟器已启动")
        print("  2. frida-server 已运行 (adb shell /data/local/tmp/frida-server -l 0.0.0.0:27056 &)")
        return

    # 尝试附加游戏进程
    print("查找游戏进程...")
    app = None
    try:
        app = dev.get_frontmost_application()
        print(f"  前台应用: {app.name} (PID={app.pid})")
    except:
        pass

    if not app or "proj.xqj" not in getattr(app, 'identifier', ''):
        print("尝试枚举进程找 proj.xqj ...")
        try:
            for proc in dev.enumerate_processes():
                if "xqj" in proc.name.lower() or "xunqin" in proc.name.lower():
                    print(f"  找到: {proc.name} (PID={proc.pid})")
                    app = proc
                    break
        except:
            pass

    if not app:
        pid = input("请输入游戏PID (或直接回车尝试5630): ").strip()
        if not pid:
            pid = 5630
        else:
            pid = int(pid)
    else:
        pid = app.pid

    print(f"附加 PID={pid} ...")
    try:
        session = dev.attach(pid)
    except Exception as e:
        print(f"[失败] 无法附加: {e}")
        return

    script = session.create_script(JS)
    script.on("message", on_msg)
    script.load()
    time.sleep(1)

    print("""
┌─────────────────────────────────────────────────────────────┐
│                                                             │
│  【现在请操作游戏】                                          │
│                                                             │
│  点击任意一个传送门，走过去触发传送。                         │
│                                                             │
│  观察本窗口输出:                                             │
│    - handler_enter → 确认handler被调用                       │
│    - handler_url  → 确认URL格式                              │
│    - portal_packet → 确认传送包内容                          │
│                                                             │
│  走完后回来，输入命令进行下一步测试。                         │
│                                                             │
└─────────────────────────────────────────────────────────────┘
""")

    # 交互命令
    print("命令: status / last / try <URL> / tryfull <URL> / quit\n")

    try:
        while True:
            cmd = input("(handler)> ").strip()
            if not cmd:
                continue

            if cmd == "quit" or cmd == "exit":
                break

            elif cmd == "status":
                try:
                    r = script.exports_sync.get_status()
                    print(json.dumps(json.loads(r), indent=2, ensure_ascii=False))
                except Exception as e:
                    print(f"Error: {e}")

            elif cmd == "last":
                try:
                    r = script.exports_sync.get_last_call()
                    print(json.dumps(json.loads(r), indent=2, ensure_ascii=False))
                except Exception as e:
                    print(f"Error: {e}")

            elif cmd == "calls":
                try:
                    r = script.exports_sync.get_all_calls()
                    print(json.dumps(json.loads(r), indent=2, ensure_ascii=False))
                except Exception as e:
                    print(f"Error: {e}")

            elif cmd == "key":
                try:
                    k = script.exports_sync.get_key()
                    print(f"XOR key = 0x{k:02x}" if k else "key not detected yet")
                except Exception as e:
                    print(f"Error: {e}")

            elif cmd.startswith("try "):
                url = cmd[4:].strip()
                print(f"尝试调用 handler('{url}') ...")
                print("警告: 可能会掉线或崩溃，确定吗? (yes/no)")
                ok = input("> ").strip().lower()
                if ok == "yes":
                    try:
                        r = script.exports_sync.try_call_handler(url)
                        print(f"结果: {r}")
                    except Exception as e:
                        print(f"Error: {e}")

            elif cmd.startswith("tryfull "):
                url = cmd[7:].strip()
                print(f"尝试完整调用 func1+handler('{url}') ...")
                print("警告: 可能会掉线或崩溃，确定吗? (yes/no)")
                ok = input("> ").strip().lower()
                if ok == "yes":
                    try:
                        r = script.exports_sync.try_call_full(url)
                        print(f"结果: {r}")
                    except Exception as e:
                        print(f"Error: {e}")

            elif cmd == "help":
                print("""
命令说明:
  status   - 查看当前状态 (fd, key, handler地址等)
  last     - 查看最近一次handler调用的参数
  calls    - 查看所有handler调用记录
  key      - 查看XOR加密密钥
  try <URL>     - 用上次的handlerObj直接调用handler
                  例: try xqj://map?name=邯郸行政区
  tryfull <URL> - 完整模拟: func1初始化 + handler调用
                  例: tryfull xqj://map?name=汉中行政区
  quit     - 退出
""")

            else:
                print(f"未知命令: {cmd} (输入 help 查看帮助)")

    except KeyboardInterrupt:
        print("\n中断")
    finally:
        session.detach()
        print("已断开")


if __name__ == "__main__":
    main()
